### PR TITLE
MergeSort implementation in C#

### PR DESCRIPTION
**Pull request by Omar-Hegazi**
### Have you read the [Contributing Guidelines on Pull Requests](https://github.com/TesseractCoding/NeoAlgo/blob/master/CONTRIBUTING.md#pull-requests)?

(Write your answer here.)
yes
### Description

(Write your answer here.)
I added an implementation of merge sort algorithm in C#. I build a merge function that compares the elements of an array and sort them in temporary array then place them into the original array. I build two MergeSort function the first is the one uses recursion to break the problem into smaller portions and the other is to reduce the number of parameter into one. I uset temporary array as parameter so the algorithm only create it one time.
### Checklist

- [yes] I've read the contribution guidelines.
- [yes ] I've checked the issue list before deciding what to submit.
- [ yes] I've edited the `README.md` and link to my code.

## Related Issues or Pull Requests

(Write your answer here.)
No

### CODE REVIEW

**Review by HarshCasper — CHANGES_REQUESTED** [review]
Few changes: 
- Add a beginner friendly pitch about the Code 
- Add sample I/O at the end of the Code
- Add the Complexity at the end of the Code

@gat786 You can review this as well. I don't really have any expertise in C#

File: C-Sharp/sort/MergeSort.cs [hunk 1]
```diff
@@ -0,0 +1,49 @@
+﻿class mergeSort
```
Comment (gat786) [inline C-Sharp/sort/MergeSort.cs]: make sure you write class name in CamelCase where the first letter is capital as well

**Comment by Omar-Hegazi** [comment]
I've Done the requested changes.

File: C-Sharp/sort/MergeSort.cs [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+    }
+    //Limit the number of parameters
+    public static void MergeSort(int[] arr)
+    {
+        MergeSort(arr, new int[arr.Length], 0, arr.Length - 1);
+
+    }
+}
+
```
Comment (gat786) [inline C-Sharp/sort/MergeSort.cs:90]: Give an example of how this merge sort will work 
Add the different complexity details as well such as time or space complexities

File: C-Sharp/sort/MergeSort.cs [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+            MergeSort(arr, C, l, mid);
+            MergeSort(arr, C, mid + 1, r);
+            merge(arr, C, l, mid, r);
+        }
+
+    }
+    //Limit the number of parameters
+    public static void MergeSort(int[] arr)
+    {
+        MergeSort(arr, new int[arr.Length], 0, arr.Length - 1);
```
Comment (gat786) [inline C-Sharp/sort/MergeSort.cs]: make the input array dynamic i.e. let user fill in the input array using some loops

File: C-Sharp/sort/MergeSort.cs [hunk 4]
```diff
@@ -0,0 +1,49 @@
+﻿class mergeSort
+{
+    //The merge step in the sorting problem to sort the elements
+    private static void merge(int[] arr, int[] C, int l, int mid, int r)
```
Comment (gat786) [inline C-Sharp/sort/MergeSort.cs]: Why do some functions have UpperCamelCase and some functions have lowerCamelCase convention? convert everything to UpperCamelCase

File: C-Sharp/sort/MergeSort.cs (hunk 1 above)
Comment (Omar-Hegazi) [reply C-Sharp/sort/MergeSort.cs]: Done

File: C-Sharp/sort/MergeSort.cs (hunk 2 above)
Comment (Omar-Hegazi) [reply C-Sharp/sort/MergeSort.cs:90]: Done

File: C-Sharp/sort/MergeSort.cs (hunk 3 above)
Comment (Omar-Hegazi) [reply C-Sharp/sort/MergeSort.cs]: Done, I added Main method for that specification.

File: C-Sharp/sort/MergeSort.cs (hunk 4 above)
Comment (Omar-Hegazi) [reply C-Sharp/sort/MergeSort.cs]: Done

**Comment by Omar-Hegazi** [comment]
Thank you for your time and helpful review. I've completed the requested changes. I hope you can review my last commit.

File: C-Sharp/sort/MergeSort.cs [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+// Implementation of merge sort
+
+using System;
+
+namespace MergeSortAlgorithm
+{
+    class Merge_Sort
+    {
+        //The merge step in the sorting problem to sort the elements
+        private static void Merge(int[] arr, int[] C, int l, int mid, int r)
```
Comment (gat786) [inline C-Sharp/sort/MergeSort.cs]: Instead of l and r use meaningful variable names. It doesn't change anything but makes code more readable.

Comment (gat786) [reply C-Sharp/sort/MergeSort.cs]: You have written the code so it makes sense to you but when you put it out to the public it is more readable if variable names make sense

Comment (gat786) [reply C-Sharp/sort/MergeSort.cs]: This whole merge function contains confusing variable names. Although it might work its a little confusing to me.

Comment (gat786) [reply C-Sharp/sort/MergeSort.cs]: Everything else looks great

**Review by gat786 — CHANGES_REQUESTED** [review]

File: C-Sharp/sort/MergeSort.cs (hunk 5 above)
Comment (Omar-Hegazi) [reply C-Sharp/sort/MergeSort.cs]: Thank you for your time and helpful review. I've fixed the variables names to more readable form.  I hope you can review my last commit.

**Review by gat786 — APPROVED** [review]

**Review by gat786 — APPROVED** [review]

**Comment by gat786** [comment]
Looks good to me now. @HarshCasper do the deed 🎊

File: C-Sharp/sort/MergeSort.cs [hunk 6]
```diff
@@ -0,0 +1,95 @@
+﻿// Merge sort is an efficient, general-purpose, comparison-based sorting algorithm.
+// Implementation of merge sort
```
Comment (HarshCasper) [inline C-Sharp/sort/MergeSort.cs]: Add a better description

**Review by HarshCasper — CHANGES_REQUESTED** [review]
Some of the comments are redundant. Remove them

File: C-Sharp/sort/MergeSort.cs (hunk 6 above)
Comment (Omar-Hegazi) [reply C-Sharp/sort/MergeSort.cs]: Done

Comment (Omar-Hegazi) [reply C-Sharp/sort/MergeSort.cs]: I only changed the comments.  but I got this error in checks "DeepSource: Java (beta) Failing after 7s — Analysis Failed: Issues detected that need attention."

**Review by gat786 — APPROVED** [review]

**Review by HarshCasper — APPROVED** [review]

**Comment by gat786** [comment]
There you go :-)

**Pull request merged**